Sources/ServiceLifecycle/ServiceRunner.swift

```diff
@@ -0,0 +1,265 @@
+            // since we want to signal them individually and wait for a single service
+            // to finish before moving to the next one
+            var gracefulShutdownManagers = [GracefulShutdownManager]()
+            gracefulShutdownManagers.reserveCapacity(self.services.count)
+
+            for (index, service) in self.services.enumerated() {
+                self.logger.debug(
+                    "Starting service",
+                    metadata: [
+                        self.configuration.logging.serviceKey: "\(service)",
+                    ]
+                )
+
+                let gracefulShutdownManager = GracefulShutdownManager()
+                gracefulShutdownManagers.append(gracefulShutdownManager)
+
+                group.addTask {
+                    return await TaskLocals.$gracefulShutdownManager.withValue(gracefulShutdownManager) {
+                        do {
+                            try await service.run()
+                            return .serviceFinished(service: service, index: index)
+                        } catch {
+                            return .serviceThrew(service: service, index: index, error: error)
+                        }
+                    }
+                }
+            }
+
+            precondition(gracefulShutdownManagers.count == self.services.count, "We did not create a graceful shutdown manager per service")
+
+            // We are going to wait for any of the services to finish or
+            // the signal sequence to throw an error.
+            while !group.isEmpty {
+                // No child task is actually throwing here so the try! is safe
+                let result: ChildTaskResult? = await group.next()
+
+                switch result {
+                case .serviceFinished(let service, _):
+                    // If a long running service finishes early we treat this as an unexpected
+                    // early exit and have to cancel the rest of the services.
+                    self.logger.error(
+                        "Service finished unexpectedly. Cancelling all other services now",
+                        metadata: [
+                            self.configuration.logging.serviceKey: "\(service)",
+                        ]
+                    )
+
+                    group.cancelAll()
+                    return .failure(ServiceRunnerError.serviceFinishedUnexpectedly())
+
+                case .serviceThrew(let service, _, let error):
+                    // One of the servers threw an error. We have to cancel everything else now.
+                    self.logger.error(
+                        "Service threw error. Cancelling all other services now",
+                        metadata: [
+                            self.configuration.logging.serviceKey: "\(service)",
+                            self.configuration.logging.errorKey: "\(error)",
+                        ]
+                    )
```

Comment:
how do we think about using the error log level here? In an sswg logging doc we say, that normally libs should not log higher than debug.

Reply:
I changed the logging to `debug` for everything now. We could make this configurable in the future. Any error that happens is thrown from the `run()` method anyhow so users could catch and log it.

Reply:
Right, I think that's more a general rule. If the error is returned _anyway_, don't log it (in non-debug levels) because the user can if they choose too.

It's a little more difficult with things that are recovered in the bowels of libraries and are not returned to the user, usually when things get retried. Think some HTTP client which fails to open a TCP connection on the first try, succeeding on the second. The user will not see an error because it could perform its job after all.
But what if, say, 50% of all TCP connections fail to establish then you likely want to know about it (despite the HTTP client being able to handle it internally by retrying). Personally, I think that can usually be done by sampling requests and having each 10,000th request or so log everything at `debug` but that requires the `Logger` to be passed down to everything or having [this](https://github.com/apple/swift-log/issues/261) work.

Reply:
Yeah it is a fair point and I think a discussion orthogonal to service-lifecycle. I followed the logging guidelines here now.

From my recent work on other libraries, I personally find logging always at debug/trace level in libraries to be limiting because as you said. There are sometimes cases in libraries where you hit unexpected behaviours but handle them gracefully. Logging those at debug level means that they basically never surface unless they cause business logic faults. The library could provide an error delegate for something like this but IMO it would be more straight forward to log those at a higher level so they surface and be reported back as bugs.

Though let's move that discussion over to logging :D

Reply:
I'd say the user really should be in control. If you _do_ want to see the logs from a particular library, you can pass a logger with a `.debug` log level to the library. Or you can select that library using the `source` field or so.

Regardless, "basically never surface[s]" is only true if you don't sample the requests at all. Sampling over time is really powerful because say you have a failure that only happens in 1% of the cases, then you'd expect to see it once after 100 requests. If you sampled every 10,000th request, you'd expect to see it in your logs after 1M requests. For services that have significant load, this would still mean that you'd see it frequently enough. If that frequency is too low for you, maybe sample every 1,000th request. In general: Sample as high as you can without violating your latency SLOs and breaking the bank with emitting too many logs :). If you have a service that handles only like 100 requests per second, you might get a way with just logging everything at debug, depending on your latency & money (for log storage/retention) budget.